privacy/activitycontrol.go

```diff
@@ -103,16 +103,42 @@ func activityRulesToEnforcementRules(rules []config.ActivityRule) ([]Rule, error
 	return enfRules, nil
 }
 
-func conditionToRuleComponentNames(conditions []string) ([]Component, error) {
-	sn := make([]Component, 0)
-	for _, condition := range conditions {
-		scope, err := ParseComponent(condition)
+func conditionToRuleComponentNames(conditions config.ActivityCondition) ([]Component, error) {
+	// condition can contain more than one component type and component name may have more than one names
+	// in this case scopes should have all combinations of names and types
+	// for instance for the config like this
+	//     "condition": {
+	//          "componentName": ["enabledAnalytics", "filelogger", "bidder.appnexus"],
+	//           "componentType":["analytics", "rtd"]
+	//     }
+	// result scopes should be:
+	// analytics.enabledAnalytics
+	// analytics.filelogger
+	// rtd.enabledAnalytics
+	// rtd.filelogger
+	// bidder.appnexus
+
+	components := make([]Component, 0)
+	for _, conditionName := range conditions.ComponentName {
+		component, err := ParseComponent(conditionName)
 		if err != nil {
-			return nil, err
+			return components, err
+		}
+		if component.Type != "" {
+			components = append(components, component)
+		} else if component.Type == "" && len(conditions.ComponentType) == 0 {
+			component.Type = ComponentTypeBidder
+			components = append(components, component)
+		} else {
+			for _, conditionType := range conditions.ComponentType {
+				if conditionType != component.Type {
+					newSN := Component{Type: conditionType, Name: component.Name}
+					components = append(components, newSN)
+				}
+			}
 		}
-		sn = append(sn, scope)
 	}
-	return sn, nil
+	return components, nil
 }
```

Comment:
Why are you combining the `ComponentName` and `ComponentType` fields here? They are handled separately, so this code just needs to parse `ComponentName`. Right?

Reply:
Yes, correct. This is a function to parse ComponentName only. We set component Type to bidder only in case if component name has this format: "bidder.componentName". 
If we have more than one component name we return a list of components with component.Name only (if bidder is not specified) and component type will be added later. I added a comment with the description and example of what will happen if there are multiple component names and types. 

```go
        // condition can contain more than one component type and component name may have more than one names
	// in this case scopes should have all combinations of names and types
	// for instance for the config like this
	//     "condition": {
	//          "componentName": ["enabledAnalytics", "filelogger", "bidder.appnexus"],
	//           "componentType":["analytics", "rtd"]
	//     }
	// result scopes should be:
	// analytics.enabledAnalytics
	// analytics.filelogger
	// rtd.enabledAnalytics
	// rtd.filelogger
	// bidder.appnexus
```

I found that before this fix in case if we have a config like this:
```json
"condition": {
	//          "componentName": [ "filelogger"],
	//           "componentType":["analytics"]
	//     }
```
result component object has componentType="bidder".

Reply:
It looks like we need to incorporate componentType here too. 
I added a comment with the description and example of what will happen if there are multiple component names and types.

        // condition can contain more than one component type and component name may have more than one names
	// in this case scopes should have all combinations of names and types
	// for instance for the config like this
	//     "condition": {
	//          "componentName": ["enabledAnalytics", "filelogger", "bidder.appnexus"],
	//           "componentType":["analytics", "rtd"]
	//     }
	// result scopes should be:
	// analytics.enabledAnalytics
	// analytics.filelogger
	// rtd.enabledAnalytics
	// rtd.filelogger
	// bidder.appnexus
I found that before this fix in case if we have a config like this:

"condition": {
        "componentName": [ "filelogger"],
         "componentType":["analytics"]
    }
result component object has componentType="bidder".

Should we rename this function to `conditionToRuleComponent` ?

Reply:
> result component object has componentType="bidder".

Right. That's the desired behavior. There is no guarantee of name uniqueness between adapters, analytics, and modules. The only way we know that `filelogger` is meant to be an `analytics` type is for that to be specified. Since adapters are most common, we made that the default for convience.

`"componentName": [ "filelogger"]` should be interpreted as `bidder.filelogger`.

> in this case scopes should have all combinations of names and types

That's not my expectation and I don't believe that's in the spec.

Reply:
This is so confusing. My suggestion may not be right, it's easy to revert it back. 
If I refer to the issue https://github.com/prebid/prebid-server/issues/2591 it says:
```
The following attributes may be seen:

componentName - this is the biddercode, the analytics adapter code, or the module code.
componentType - this is "bidder", "analytics", or for modules: if the module tag contains "rtd", then type is "rtd", otherwise type is "general".

```
In examples below it shows:
```
 rules: [{
          condition: {
            componentType: ["bidder", "analytics"]
          },
          allow: false
        }
```
This gave me the idea that config name can only have name and config type can only have a type and we will need to combine them together.

Reply:
Interesting. Reviewing the history of issue #2591, seems the dot-notation was removed the first day of the focus group. I didn't remember that change. It does offer a lot of simplification with the assumption that name conflicts between analytics, adapters, and modules are rare.

In this case, let's keep the target as a `Component`, but change the `componentName` constraint to simple an array of strings. We can delete the complex `Component` comparison logic and for `componentName` ignore the type and for `componentType` ignore the name. We don't need to compute the matrix of all possibilities. These comparisons should be case insensitive.

Reply:
That would be a lot of tangental work to include in this PR. I put up https://github.com/prebid/prebid-server/pull/3081 for the change.